extension-manifest-v3/src/background/utils/engines.js

```diff
@@ -0,0 +1,336 @@
+/**
+ * Ghostery Browser Extension
+ * https://www.ghostery.com/
+ *
+ * Copyright 2017-present Ghostery GmbH. All rights reserved.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0
+ */
+
+import * as IDB from 'idb';
+import {
+  FiltersEngine,
+  ENGINE_VERSION,
+  getLinesWithFilters,
+  mergeDiffs,
+} from '@cliqz/adblocker';
+
+import { registerDatabase } from '/utils/indexeddb.js';
+
+const engines = new Map();
+
+function loadFromMemory(name) {
+  return engines.get(name);
+}
+
+function saveToMemory(name, engine) {
+  engines.set(name, engine);
+}
+
+const DB_NAME = registerDatabase('engines');
+
+async function getDB() {
+  if (!getDB.current) {
+    getDB.current = IDB.openDB(DB_NAME, 1, {
+      upgrade(db) {
+        db.createObjectStore('engines');
+      },
+      async blocking() {
+        const db = await getDB.current;
+        db.close();
+        getDB.current = null;
+      },
+    });
+  }
+
+  return getDB.current;
+}
+
+async function loadFromStorage(name) {
+  try {
+    const db = await getDB();
+
+    const tx = db.transaction('engines');
+    const table = tx.objectStore('engines');
+
+    const engineBytes = await table.get(name);
+
+    if (engineBytes) {
+      const engine = FiltersEngine.deserialize(engineBytes);
+      saveToMemory(name, engine);
+
+      return engine;
+    }
+  } catch (e) {
+    console.error(`Failed to load engine "${name}" from storage`, e);
+  }
+
+  return null;
+}
+
+async function saveToStorage(name) {
+  try {
+    const engine = loadFromMemory(name);
+    const db = await getDB();
+
+    const tx = db.transaction('engines', 'readwrite');
+    const table = tx.objectStore('engines');
+
+    await table.put(engine.serialize(), name);
+  } catch (e) {
+    console.error(`Failed to save engine "${name}" to storage`, e);
+  }
+}
+
+function check(response) {
+  if (!response.ok) {
+    throw new Error(
+      `Failed to fetch engine "${name}": ${response.status}: ${response.statusText}`,
+    );
+  }
+
+  return response;
+}
+
+async function update(name) {
+  try {
+    const urlName =
+      name === 'trackerdb'
+        ? 'trackerdbMv3'
+        : `dnr${__PLATFORM__ === 'firefox' ? '' : '-cosmetics'}-${name}`;
+
+    const data = await fetch(
```

Comment:
Since this polls for updates, caching can create problems since you can end up with an old version. If the update call here is guaranteed to be called not too often, you can disabling it with

```
fetch(url, {
  method: 'GET',
  cache: 'no-cache',
  credentials: 'omit',
}
```
  
(e.g. https://github.com/whotracksme/webextension-packages/blob/257b4f6dba519497119ecf9d16c1475ef435ff49/packages/reporting/src/patterns-updater.js#L193)

But if it is called frequently, disabling caching can be wasteful for the client. One compromise would be to introduce a "force" flag and only bypass the browser cache in force mode.

Reply:
This is a service, which is in our control. Why should we bypass our own cache settings? The update process is triggered once a day, so if the endpoint has an hour of cache, it should be fine. The only case is when you want to trigger the update process manually, but if we need to have the ability to fetch it more often, let's change the cache setting on the server side - clients normally don't call it more often than once a day (so in that case, the cache could be turned off completely).

Reply:
Regarding the server settings: serving a mutable file as "immutable" (even with only one hour) is indeed a bit contradicting. A release of the adblocker invalidates the CDN, which makes the new content accessible from the internet.

For the client: The interesting use case is indeed to force an update. But I think, we can stick with the current workflow, which is to install a fresh extension and test it with that. That is the only reliable way to force the latest block list (also for the current MV2 Ghostery).

Though we should keep in mind that forcing an update through the dev UI is unreliable.